teps/0130-pipeline-level-service.md

```diff
@@ -0,0 +1,759 @@
+
+#### Docker builds
+
+The docker-build catalog Task only allows building a single image and pushing it to a registry.
+The following use cases are not addressed:
+- Building multiple images concurrently with the same daemon to save resources and share cached base layers
+  - For example, fanning out image build Tasks for the same repository via a Matrix
+- Introducing additional steps, such as testing, in between a build and a push
+  - A user may also want to guard guard execution of the push based on the results of those tasks
+  - A user may also want to use the built image in these subsequent tasks without having to first push to and pull from a registry;
+  for example, if using `docker exec` or `docker-compose`
+
+Allowing multiple Tasks in a PipelineRun to use the same docker daemon allows PipelineRuns with
+docker builds to take advantage of Pipeline features (e.g. using matrix to build multiple images,
+executing a push conditionally on previous steps) while still having access to the same local
+cache of built images and avoid having to push to a registry in between.
+
+#### Other uses for this feature
+
+A few other uses for Pipeline-level services have been proposed, but Docker builds are the highest priority use case identified so far.
+
+- **Integration tests**: This feature may also be useful for integration tests that use a service to communicate with a test database.
+However, integration tests are likely limited to a single Task; i.e. it's unlikely that a user would want multiple integration testing Tasks
+to share the same test database. A test database Pipeline-level sidecar service could be limited to one Task (or a subset of Tasks) using
+[Pipelines in Pipelines](./0056-pipelines-in-pipelines.md), or this use case could be met by allowing Pipelines or TaskRuns
+to specify Task-level sidecars, as proposed in
+[TEP-0126: Allow Task sidecars to be specified in PipelineRun](https://github.com/tektoncd/community/pull/877).
+
+- **Heartbeats**: The [original feature request](https://github.com/tektoncd/pipeline/issues/2973) for Pipeline-level sidecar services was within a Pipeline
+that uses Boskos resources. The first Task acquires the resources and creates a pod to send heartbeats to Boskos;
+intermediate Tasks use the Boskos resources, and a Finally Task would release the lease on the resources and clean up the pod sending heartbeats.
+One user [suggested](https://github.com/tektoncd/pipeline/issues/2973#issuecomment-672152635) that Boskos usage should be restricted to a subset of Tasks,
+which could be done with Pipelines in Pipelines.
+
+### Goals
+
+- More flexible Docker build Pipelines
+
+### Requirements
+
+- A Docker image built during a Pipeline can be used in a subsequent Task without first pushing it to an image registry
+- Service is only accessible to the single PipelineRun that uses it
+(doing otherwise would violate [SLSA L3 build isolation requirements](https://slsa.dev/spec/v0.1/requirements#isolated))
```

Comment:
Noted-- I made the goal that we should support SLSA L3 and added a "security" section noting some of the points on docker builds.

Reply:
Also one point on SLSA. I feel tektoncd should help setuping up CI system that are SLSA compliant (L1 -> LX), but tektoncd itself doesn't have be compliant or to enforce anything related to SLSA (as it can be use for something completely different, like data science with kubeflow, …)